### PR TITLE
PP-7709 Allow products to be retrieved by account ID and type

### PR DESCRIPTION
**Pull request by alexbishop1**
Enhance the `/gateway-account/{gatewayAccountId}/products` endpoint so that it takes an optional `type` parameter with a single value like `ADHOC` or `PROTOTYPE` and filters the results appropriately.

### CODE REVIEW

File: src/main/java/uk/gov/pay/products/service/ProductFinder.java [hunk 1]
```diff
@@ -92,26 +95,31 @@ public Boolean deleteByGatewayAccountIdAndExternalId(Integer gatewayAccountId, S
         return productDao.findByGatewayAccountId(gatewayAccountId)
                 .stream()
                 .map(ProductEntity::toProduct)
-                .collect(Collectors.toList())
+                .map(linksDecorator::decorate)
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    @Transactional
+    public List<Product> findByGatewayAccountIdAndType(Integer gatewayAccountId, ProductType type) {
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/products/service/ProductFinder.java:103]: Personally I'd prefer most methods in ProductFinder to be simply `find` with overloaded methods. But it's not a biggie.

File: src/main/java/uk/gov/pay/products/service/ProductFinder.java [hunk 2]
```diff
@@ -92,26 +95,31 @@ public Boolean deleteByGatewayAccountIdAndExternalId(Integer gatewayAccountId, S
...
     @Transactional
     public List<ProductUsageStat> findProductsAndUsage(Integer gatewayAccountId) {
         return productDao.findProductsAndUsage(gatewayAccountId)
                 .stream()
-                .map(productUsageStat -> {
-                    productUsageStat.setProduct(linksDecorator.decorate(productUsageStat.getProduct()));
-                    return productUsageStat;
-                })
-                .collect(Collectors.toList());
+                .peek(productUsageStat -> productUsageStat.setProduct(linksDecorator.decorate(productUsageStat.getProduct())))
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/products/service/ProductFinder.java:122]: nice

File: src/test/java/uk/gov/pay/products/resources/ProductResourceIT.java [hunk 3]
```diff
@@ -942,6 +989,28 @@ public void findProductsByGatewayAccountId_shouldNotReturnInactiveProducts() {
                 .body("", hasSize(0));
     }
 
+    @Test
+    public void findProductsByGatewayAccountIdWithType_shouldNotReturnInactiveProducts() {
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/products/resources/ProductResourceIT.java:993]: I feel this might be unnecessary as it's already tested in ProductDaoIT

**Review by oswaldquek — APPROVED** [review]
Some comments but nothing to stop it from being approved.

**Comment by oswaldquek** [comment]
It would've been great to explain how this PR fits the story description: "Add a mechanism in the admin tool for a user with the agent-initiated-moto-payments:initiate to initiate a payment from a AGENT_INITIATED_MOTO product."

File: src/main/java/uk/gov/pay/products/service/ProductFinder.java (hunk 1 above)
Comment (alexbishop1) [reply src/main/java/uk/gov/pay/products/service/ProductFinder.java:103]: I was just following the existing conventions.

File: src/main/java/uk/gov/pay/products/service/ProductFinder.java (hunk 2 above)
Comment (alexbishop1) [reply src/main/java/uk/gov/pay/products/service/ProductFinder.java:122]: You can thank IntelliJ’s suggestions for that!

File: src/test/java/uk/gov/pay/products/resources/ProductResourceIT.java (hunk 3 above)
Comment (alexbishop1) [reply src/test/java/uk/gov/pay/products/resources/ProductResourceIT.java:993]: It’s to mirror the tests we have for using the endpoint without the `type` parameter. I suppose it does check that neither `ProductFinder` nor `ProductResource` blow up when they get an empty list from the DAO.

**Comment by alexbishop1** [comment]
One of the things we’re going to need for the story is a way to get all products that are `AGENT_INITIATED_MOTO` and this makes that more efficient.

**Pull request merged**